File: 2012/J122.cpp
#include <cstdio>

int main() {
  int R,C,r,c;
  scanf("%d %d %d %d", &R, &C, &r, &c);
  if ((R * C % 2 == 0) || ((r - c) % 2 != 0)) {
    printf("Impossible\n");
    return 0;
  }
  for (int i=1; i<r; i++) {
    for (int j=0; j<C-1; j++)
      printf("%c", 'a' + (i % 2) + (j & 2));
    printf("%c\n", 'm' + ((i-1) & 2));
  }
  if (r % 2 == 1) {
    for (int j=0; j<c-1; j++)
      printf("%c", 'e' + (j & 2));
    printf("*");
    for (int j=c; j<C; j++)
      printf("%c", 'e' + ((j+1) & 2));
    printf("\n");
  } else {
    printf("%c", 'p' + (r & 2));
    for (int j=0; j<c-2; j++)
      printf("%c", 'e' + (j & 2));
    printf("*");
    for (int j=c; j<C-1; j++)
      printf("%c", 'e' + (j & 2));
    printf("%c\n", 'm' + ((r-1) & 2));
  }
  for (int i=r+1; i<=R; i++) {
    printf("%c", 'p' + (i & 2));
    for (int j=0; j<C-1; j++)
      printf("%c", 'a' + (i % 2) + (j & 2));
    printf("\n");
  }
  return 0;
}

File: 2012/J122.pas
var
  r, c, rr, cc, i, j: longint;
begin
  read(r, c, rr, cc);
  if ((r * c mod 2 = 0) or ((rr - cc) mod 2 <> 0)) then
  begin
    writeln('Impossible');
    exit
  end;
  for i := 1 to rr - 1 do
  begin
    for j := 0 to c - 2 do
      write(chr(ord('a') + (i mod 2) + (j and 2)));
    writeln(chr(ord('m') + ((i - 1) and 2)))
  end;
  if (rr mod 2 = 1) then
  begin
    for j := 0 to cc - 2 do
      write(chr(ord('e') + (j and 2)));
    write('*');
    for j := cc to c - 1 do
      write(chr(ord('e') + ((j + 1) and 2)));
    writeln
  end
  else
  begin
    write(chr(ord('p') + (rr and 2)));
    for j := 0 to cc - 3 do
      write(chr(ord('e') + (j and 2)));
    write('*');
    for j := cc to c - 2 do
      write(chr(ord('e') + (j and 2)));
    writeln(chr(ord('m') + ((rr - 1) and 2)))
  end;
  for i := rr + 1 to r do
  begin
    write(chr(ord('p') + (i and 2)));
    for j := 0 to c - 2 do
      write(chr(ord('a') + (i mod 2) + (j and 2)));
    writeln
  end;
end.
